contracts/marketplace/lib.rs

```diff
@@ -0,0 +1,299 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(min_specialization)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+pub mod marketplace {
+    use access_control::{
+        traits::AccessControlled, Role, Role::Initializer, ACCESS_CONTROL_PUBKEY,
+    };
+    use game_token::TRANSFER_FROM_SELECTOR as TRANSFER_FROM_GAME_TOKEN_SELECTOR;
+    use ink_env::{
+        call::{build_call, Call, ExecutionInput, Selector},
+        CallFlags,
+    };
+    use ink_lang::{codegen::EmitEvent, reflect::ContractEventBase};
+    use ink_prelude::{format, string::String};
+    use openbrush::contracts::psp22::PSP22Error;
+    use ticket_token::{
+        BALANCE_OF_SELECTOR as TICKET_BALANCE_SELECTOR,
+        TRANSFER_SELECTOR as TRANSFER_TICKET_SELECTOR,
+    };
+
+    use crate::marketplace::Error::MissingRole;
+
+    type Event = <Marketplace as ContractEventBase>::Type;
+
+    const DUMMY_DATA: &[u8] = &[0x0];
+
+    #[ink(storage)]
+    pub struct Marketplace {
+        total_proceeds: Balance,
+        tickets_sold: Balance,
+        min_price: Balance,
+        current_start_block: BlockNumber,
+        auction_length: BlockNumber,
+        sale_multiplier: Balance,
+        ticket_token: AccountId,
+        reward_token: AccountId,
+    }
+
+    #[derive(Clone, Eq, PartialEq, Debug, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        MissingRole(String),
```

Comment:
why not use the actual type here? MissingRole(Role)

Reply:
I guess we control this type, so we could make it Encode/Decode, etc.

Reply:
it already is afaik